spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArray.scala

```diff
@@ -66,6 +68,21 @@ case class ErrorMessageArray(errorColumnName: String = ErrorMessageArray.default
     dataFrame.withColumnIfDoesNotExist(joinToExisting)(errorColumnName, reMap(aggregatedWithoutNulls))
   }
 
+  override def errorColumnType: DataType = {
+    decomposeMap(col(errorColumnName)).expr.dataType
```

Comment:
I actually thought that this is a good idea, at first, but I played with it on Spark Shell and I don't think that what you are trying to achieve here is possible. @benedeki please feel free to disagree.

The reason is that you are creating a new Column object here, a generic one, not associated with any dataframe so no internal structure / metadata will be associated with it. So you are only creating a placeholder for a column. 

The `decomposeMap(col(errorColumnName))` make sense, but it will produce a Column, not a dataType. Then, `expr` will produce a general expression full of NamePlaceholder elements, and then `.dataType` call will fail on: `org.apache.spark.SparkUnsupportedOperationException: Cannot evaluate expression: NamePlaceholder`

These are deeper Spark internals and I don't know if you can hack it so that you get a schema in `dataType` out of it.

---

Why not to take a simpler approach? The structure is already defined above. Check how the err column is being added into the DataFrame, line:

`dataFrame.withColumn(errorColName, reMap(array_union(deMap(col(errorColName)), colToUnion)))`

or `dataFrame.withColumnIfDoesNotExist(joinToExisting)(errorColumnName, reMap(aggregatedWithoutNulls))` and try to understand the internals of functions involved in these computations.

Basically it's in the `recomposeMap` - maybe you can somehow extract & reuse that, but maybe it would be good enough to just code the structure manually into your `errorColumnType` according to this

Reply:
Btw, in terms of the schema that you are looking for, every concrete implementation should know what schema the err column will be used. See https://github.com/AbsaOSS/spark-commons/pull/103, the documentation is I think outstanding. The function `createErrorAsColumn` is responsible for creating a error Column expression. If you check what's inside, it is using `transformErrorSubmitToColumn`. It might be a good idea to read the documentation carefully and confirm your understanding with concrete implementations and unit tests. 

For example, `ErrorMessageArray` is extending `ErrorHandling` and `TransformIntoErrorMessage`. In the `TransformIntoErrorMessage`, you can see the definition of `transformErrorSubmitToColumn` that I mentioned above - this function is important for you - and the schema is defined there :) 

David's changes simplify things. You might be a bit confused about the function naming, the master branch that you are using still has `putErrorToColumn` (not transformErrorSubmitToColumn) and `evaluate` function (instead of `transformErrorSubmitToColumn`) but the concept is the same. So in your case `evaluate` function contains the err Column structure.

Reply:
Okay, thank you  Ladislav. This is crucial and insightful, and I will be sure to ask if there is something that confuses me.